python/benchmark/gen_data_distributed.py

```diff
@@ -232,6 +234,155 @@ def make_matrix_udf(iter: Iterable[pd.DataFrame]) -> Iterable[pd.DataFrame]:
+        )
+        # must replace the None to the correct type
+        params["effective_rank"] = int
+        params["random_state"] = int
+        return params
+
+    def gen_dataframe_and_meta(
+        self, spark: SparkSession
+    ) -> Tuple[DataFrame, List[str], np.ndarray]:
+        dtype = self.dtype
+        params = self.extra_params
+
+        if "random_state" not in params:
+            # for reproducible dataset.
+            params["random_state"] = 1
+
+        print(f"Passing {params} to make_regression")
+
+        rows = self.num_rows
+        cols = self.num_cols
+        assert self.args is not None
+        num_partitions = self.args.output_num_files
+
+        # Set num_partitions to Spark's default if output_num_files is not provided.
+        if num_partitions is None:
+            num_partitions = spark.sparkContext.defaultParallelism
+
+        # Retrieve input params or set to defaults.
+        seed = params["random_state"]
+        generator = np.random.RandomState(seed)
+        bias = params.get("bias", 0.0)
+        noise = params.get("noise", 0.0)
+        shuffle = params.get("shuffle", True)
+        effective_rank = params.get("effective_rank", None)
+        n_informative = params.get("n_informative", 10)
+        n_targets = params.get("n_targets", 1)
+
+        # Description (from sklearn):
+        #
+        # Input set is either well conditioned (default) or has a low rank fat tail singular profile (see LowRankMatrixDataGen).
+        # Output is generated by applying a (potentially biased) random linear regression model to the input, with n_informative
+        # nonzero regressors and some gaussian centered noise with adjustable scale.
+        #
+        # Code adapted from sklearn.datasets.make_regression().
+        if effective_rank is not None:
+            tail_strength = params.get("tail_strength", 0.5)
+            lrm_input_args = [
+                "--num_rows",
+                str(rows),
+                "--num_cols",
+                str(cols),
+                "--dtype",
+                str(dtype),
+                "--output_dir",
+                "temp",
+                "--output_num_files",
+                str(num_partitions),
+                "--effective_rank",
+                str(effective_rank),
+                "--tail_strength",
+                str(tail_strength),
+                "--random_state",
+                str(seed),
+            ]
+            # Generate a low-rank, fat tail input set.
+            X, _ = LowRankMatrixDataGen(lrm_input_args).gen_dataframe(spark)
+            assert X.rdd.getNumPartitions() == num_partitions, (
+                f"Unexpected num partitions received from LowRankMatrix;"
+                f"expected {num_partitions}, got {X.rdd.getNumPartitions()}"
+            )
+        else:
+            # Randomly generate a well-conditioned input set.
+            X = spark.createDataFrame(
```

Comment:
Will this be at a different variance than X from low_rank_matrix?

Reply:
I believe it should be, since the variance from an X in low_rank_matrix will depend on the magnitudes of the singular values, which will vary by the degree of tail_strength.

Reply:
Sorry, I meant in terms of how the variance scales/depends on the number of examples generated.